lib/artefact_retriever_with_request_cache.rb

```diff
@@ -0,0 +1,51 @@
+class RequestMissingException < NotImplementedError; end
+
+class ArtefactRetrieverWithRequestCache
+  def initialize(artefact_retriever: ArtefactRetrieverFactory.artefact_retriever)
+    @artefact_retriever = artefact_retriever
+  end
+
+  def set_request(request)
+    @request = request
+  end
+
+  def fetch_artefact(slug, edition = nil)
+    raise RequestMissingException if @request.nil?
+    return false if error?
+
+    begin
+      cache_artefact slug, edition
+    rescue => err
+      cache_error err
+      false
+    end
+  end
+
+  def cache_artefact(slug, edition)
```

Comment:
`cache_artefact` and `cache_error` could be private methods since they're only called from within this class and because we don't have unit tests for these methods?